packages/langium/src/lsp/langium-grammar-hover-provider.ts

```diff
@@ -0,0 +1,74 @@
+/******************************************************************************
+ * Copyright 2021 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+import { Hover, HoverParams } from 'vscode-languageserver';
+import {CrossReference, findLeafNodeAtOffset, findNameAssignment, Grammar, isCrossReference, isParserRule, isType, LangiumDocuments, LangiumServices, RuleCall } from '..';
+import { AstNode } from '../syntax-tree';
+import { findRelevantNode } from '../utils/cst-util';
+import { MaybePromise } from '../utils/promise-util';
+import { LangiumDocument } from '../workspace/documents';
+import { MultilineCommentHoverProvider } from './hover-provider';
+
+export class LangiumGrammarHoverProvider extends MultilineCommentHoverProvider {
+
+    readonly grammar: Grammar;
+    readonly documents: LangiumDocuments;
+    constructor(services: LangiumServices) {
+        super(services);
+        this.grammar = services.Grammar;
+        this.documents = services.shared.workspace.LangiumDocuments;
+    }
+
+    getHoverContent(document: LangiumDocument<AstNode>, params: HoverParams): MaybePromise<Hover | undefined> {
+        const rootNode = document.parseResult?.value?.$cstNode;
+        if(rootNode)  {
+            const offset = document.textDocument.offsetAt(params.position);
+            const cstNode = findLeafNodeAtOffset(rootNode, offset);
+            if(cstNode){
+                const relevantNode = findRelevantNode(cstNode);
+                if(isCrossReference(relevantNode)){
+                    return this.getCrossReferenceHoverContent(relevantNode);
+                } else {
+                    return super.getHoverContent(document, params);
+                }
+            }
+        }
+        return undefined;
+    }
```

Comment:
Suggestion: If you're already using the `MultilineCommentHoverProvide`, you should do yourself a favor and just override the `getAstNodeHoverContent` method.

https://github.com/langium/langium/blob/244442e047424c3c519d1017a81c97225e4545ae/packages/langium/src/lsp/hover-provider.ts#L52

Something like this:

```ts
protected getAstNodeHoverContent(astNode): Hover | undefined {
  if (isCrossReference(astNode) {
    return this.getCrossReferenceHoverContent(astNode);
  }
  return super.getAstNodeHoverContent(astNode);
}
```